tests/test_components.py

```diff
@@ -0,0 +1,136 @@
+        if "packagename" in template_vars:
+            package = template_vars["packagename"]
+        else:
+            package = template_vars["servicename"]
+        component = package_to_components.get(package, package)
+        reason = (
+            "rule uses template '%s' checking service '%s' provided by "
+            "package '%s' which is a package that already belongs to "
+            "component '%s'" % (
+                template_name,  template_vars["servicename"],
+                package, component))
+        components.append((component, reason))
+    return components
+
+
+def test_nonexistent_rules(rules_in_benchmark, rules_with_component):
+    nonexistent_rules = rules_with_component - rules_in_benchmark
+    if nonexistent_rules:
+        print("The following rules aren't part of the benchmark:")
+        for rule_id in nonexistent_rules:
+            print("- %s" % (rule_id))
+        return False
+    return True
+
+
+def test_unmapped_rules(rules_in_benchmark, rules_with_component):
+    unmapped_rules = rules_in_benchmark - rules_with_component
+    if unmapped_rules:
+        print("The following rules aren't part of any component:")
+        for x in unmapped_rules:
+            print("- " + x)
+            return False
+    return True
+
+
+def iterate_over_all_rules(base_dir, env_yaml):
+    """
+    Generator which yields all rule objects within a given base_dir,
+    recursively
+    """
+    for rule_dir in ssg.rules.find_rule_dirs(base_dir):
+        rule_yaml_file_path = ssg.rules.get_rule_dir_yaml(rule_dir)
+        try:
+            rule = ssg.build_yaml.Rule.from_yaml(rule_yaml_file_path, env_yaml)
+        except ssg.yaml.DocumentationNotComplete:
+            pass
+        yield rule
+
+
+def test_templates(
+        linux_os_guide_dir, env_yaml,
+        package_to_components, rules_to_components, template_to_component):
+    result = True
+    for rule in iterate_over_all_rules(linux_os_guide_dir, env_yaml):
+        candidates = get_components_by_template(
+            rule, package_to_components, template_to_component)
+        rule_components = [c.name for c in rules_to_components[rule.id_]]
+        for candidate, reason in candidates:
+            if candidate not in rule_components:
+                result = False
+                print(
+                    "Rule '%s' should be assigned to component '%s', "
+                    "because %s." % (rule.id_, candidate, reason))
+    return result
+
+
+def main():
+    result = 0
+    args = parse_args()
+    components_dir = os.path.join(args.source_dir, "components")
+    components = ssg.components.load(components_dir)
+    rule_to_components = ssg.components.rule_components_mapping(components)
+    linux_os_guide_dir = os.path.join(args.source_dir, "linux_os", "guide")
+    rules_with_component = set(rule_to_components.keys())
+    rules_in_benchmark = set(ssg.rules.find_all_rules(linux_os_guide_dir))
```

Comment:
Iterating over built rules is probably better, because it opens doors to examining remediations and checks as well, without having to deal with macros. In-source rule contents are unreliable, as there can be important unexpanded macros.
Moreover, you already use the product context in the code, so it is just one step from using the compiled contents.